feat(hardware_control): Added firmware update mechanism for submodules

vegano1 opened the pull request:
# Overview

We need to be able to update the firmware on attached submodules when the system starts up, this pr works off [RCORE-475](https://opentrons.atlassian.net/browse/RCORE-475), and [RCORE-476](https://opentrons.atlassian.net/browse/RCORE-476) to make this happen.

Closes
 [RCORE-477](https://opentrons.atlassian.net/browse/RCORE-477)
[RCORE-479](https://opentrons.atlassian.net/browse/RCORE-479)
[RCORE-481](https://opentrons.atlassian.net/browse/RCORE-481)

# Test Plan

- The robot-server starts up
- The `/usr/lib/firmware/opentrons-manifest.json` file is serialized into a dict of FirmwareUpdateType and UpdateInfo.
- The DeviceInfoCache is queried from all nodes on the network and cached in the NetworkInfo class instance.
- The OT3Controller class drives and checks if any subsystem is out of date and returns a list of submodules to be updated
- If any submodules need to be updated the _update_requied flag is set to True and subsequent calls to move the motors or anything not pertaining to firmware updates will raise a `FirmwareUpdateRequired` exception.
- The firmware will start to get installed on all out-of-date modules.
- Calling OT3Controler.update_firmware with given nodes, updates said nodes, even if they are up-to-date.
- Firmware updates are NOT started by default, only when `OT3EnableFirmwareUpdates` feature flag is set.
- The robot server is unavailable until all nodes are updated.


# Changelog

- Added utils.py module that will handle utilities for firmware updates like
   - serializing the opentrons-firmware.json manifest file
   - checking for firmware updates given a list of nodes.
- Added a mechanism to OT3Controller which checks individual subsystems and determines if they need to be updated based on the `shortsha`
- Starts firmware updates for all out-of-date subsystems on the network.
- set the update_required flag based on whether updates are available or not.
- 

# Review requests

- Does this make sense? Is there anything im overlooking?


# Risk assessment

Mid, this will raise a new exception if the firmware is out of date and we attempt to move the gantry/gripper/etc so we will need to handle that.

# Todo:
- Add code coverage for new functionality



[RCORE-475]: https://opentrons.atlassian.net/browse/RCORE-475?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RCORE-476]: https://opentrons.atlassian.net/browse/RCORE-476?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RCORE-477]: https://opentrons.atlassian.net/browse/RCORE-477?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[RCORE-479]: https://opentrons.atlassian.net/browse/RCORE-479?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[RCORE-481]: https://opentrons.atlassian.net/browse/RCORE-481?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

vegano1 committed:
added device_info module to handle DeviceInfoRequest and DeviceInfoResponse messages
device_info is now cached in ot3controller._device_info_cache dict
this is updated when the network is probed which happens when the ot3 starts up and when the instruments are cached

vegano1 committed:
add update_required flag
added requires_update decorator that raises MustUpdateFirmware if update_required flag is set
added MustUpdateFirmware exception that gets raised when a sussystem requires an update.
added the requires_update decorator to methods in the OT3Controller class that send intrusive can messages.

vegano1 committed:
Merge branch 'edge' into RCORE-475_update-required_flag

vegano1 committed:
rename exception to FirmwareUpdateRequired

vegano1 committed:
wip

vegano1 committed:
Merge branch 'RCORE-475_update-required_flag' into RCORE-477_drive_update-required

vegano1 committed:
Added firmware_manifest.py to handle loading and serializing the opentrons-firmware.json file
Added UpdateInfo dataclass to hold serialized update information loaded from opentrons-firmware.json file
serialize the manifest file on start-up and keep a record in the form {FirmwareUpdateType: UpdateInfo}

vegano1 committed:
formatting/linting

vegano1 committed:
started adding mechanism to check for updates and start a list of updates using single update mechanism

vegano1 committed:
Merge branch 'edge' into RCORE-476_fw_version_cache

vegano1 committed:
formating

vegano1 committed:
Update api/src/opentrons/hardware_control/backends/ot3controller.py

Co-authored-by: Seth Foster <seth@opentrons.com>

vegano1 committed:
consolidate device_info module into network module
add NetworkInfo class to network module that is responsible for storing network_info about the nodes on the network
moved probe into NetworkInfo class so we can cache the response we get whenever probe is called

vegano1 committed:
Merge branch 'RCORE-476_fw_version_cache' of https://github.com/Opentrons/opentrons into RCORE-476_fw_version_cache

vegano1 committed:
fix probe since it was not returning all attached nodes

vegano1 committed:
lint

vegano1 committed:
formatting...

vegano1 committed:
Merge branch 'RCORE-476_fw_version_cache' into RCORE-475_update-required_flag

vegano1 committed:
Merge branch 'edge' into RCORE-476_fw_version_cache

vegano1 committed:
Merge branch 'RCORE-476_fw_version_cache' into RCORE-475_update-required_flag

vegano1 committed:
remove requires_update decorator for update_firmware since its needed to start updates.

vegano1 committed:
cache NodeId in DeviceInfoCache instead device_id int

vegano1 committed:
Merge branch 'RCORE-476_fw_version_cache' into RCORE-475_update-required_flag

vegano1 committed:
Merge branch 'RCORE-475_update-required_flag' into RCORE-477_drive_update-required

vegano1 committed:
use network_info.device_info to check for updates
fix typing...

vegano1 committed:
move decorator to outside of class

vegano1 committed:
moved exceptions out of opentrons.hardware_control.types into new file opentrons.hardware_control.errors
satisfy mypy

vegano1 committed:
Merge branch 'RCORE-475_update-required_flag' into RCORE-477_drive_update-required

vegano1 committed:
added get_update_type to convert NodeId to FirmwareUpdateType so we know what firmware to use.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,8 +206,65 @@ def fw_version(self) -> Optional[str]:
+        """
+        if self._attached_tools is None:
+            return None
+        elif "pipette" in node_id.name:
```
I have the feeling this function should be pushed down into `opentrons_hardware`, and be a function that takes the current set of node IDs and `self._attached_tools`. All the data types that the function consumes or returns are from opentrons_hardware, and the functionality is pretty straight lookup.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,8 +206,65 @@ def fw_version(self) -> Optional[str]:
+            if update_type:
+                return update_type[0]
+        else:
+            return FirmwareUpdateType.from_string(node_id.name)
```
would be much more comfortable not involving strings here anywhere and just looking this up from the node id

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,8 +206,65 @@ def fw_version(self) -> Optional[str]:
+        for node, version_cache in self._network_info.device_info.items():
+            update_type = self.get_update_type(node)
+            if update_type == FirmwareUpdateType.unknown:
+                log.error(f"Unknown firmware update type {update_type}")
```
should also log the node id this message is about

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,8 +206,65 @@ def fw_version(self) -> Optional[str]:
+                continue
+            if version_cache.shortsha != update_info.shortsha:
+                log.debug(
+                    f"Subsystem {node} requires an update '{version_cache.shortsha} != {update_info.shortsha}'"
```
let's label in the logs which is the version cache and which is from the node

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,8 +206,65 @@ def fw_version(self) -> Optional[str]:
+        else:
+            return FirmwareUpdateType.from_string(node_id.name)
+
+    def check_firmware_updates(
```
I think this, too, can be pushed down to `opentrons_hardware` for basically the same reason - its input and output types are all defined in `opentrons_hardware`. 

That would have the interesting side effect of pushing all handling of the manifest down into `opentrons_hardware`, which would be nice IMO because that way everything about the updates is in there, except for starting. It pushes `opentrons_hardware` farther toward "a collection of functionality that higher levels can pick from"

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -184,6 +188,14 @@ def __init__(self, config: OT3Config, driver: AbstractCanDriver) -> None:
+        self._firmware_update_list: Dict[DeviceInfoCache, UpdateInfo] = {}
+        self._known_firmware = load_firmware_manifest()
+        # starts subsystem updates
+        self.do_firmware_updates()
```
this won't quite do what you want - it will create a coroutine to do the updates, but never schedule it to run, and nothing would know it exists.

I also think that, in contrast to a lot of other comments in this PR that are variants of "let's push this down a level", the decision on _when_ to do updates should be pushed _up_ a level or two. I think the hardware controller, the code in `ot3api`, should have a method that says "go do the firmware updates you currently find important", and I think that method should be called very specifically here: https://github.com/Opentrons/opentrons/blob/5a224b0e4c4c0614ac1d5af2091b32219ec7579a/api/src/opentrons/__init__.py#L161

That's because anything that happens inside 
- The `OT3Controller` constructor 
- The builder function for `OT3Controller` [here](https://github.com/Opentrons/opentrons/blob/5a224b0e4c4c0614ac1d5af2091b32219ec7579a/api/src/opentrons/hardware_control/ot3api.py#L238-L262)
- The contructor for `OT3API`

actually, in the real system, occurs in this annoying separate thread [here](https://github.com/Opentrons/opentrons/blob/5a224b0e4c4c0614ac1d5af2091b32219ec7579a/api/src/opentrons/hardware_control/thread_manager.py#L177-L196) and it's _incredibly_ annoying to find and debug an error that occurs in any of them.

Also, that first place I linked in the package `__init__` method is where we do other startup behaviors, like z-homing.

sfoster1 reviewed (changes requested):
The functionality looks good (except for the not-running-the-update-coroutine) but in my opinion this is too much functionality for `OT3Controller` specifically. I think that
- The behavior in `check_firmware_updates` and `get_update_type` should be pushed down to `opentrons_hardware`, where they can be free functions that get lists of node ids or something
- The decision point about when to do a firmware update at boot should be pushed _up_ to the package initializer `opentrons.__init__`, which pulls it out of a specific really annoying point to debug and makes it more like other startup behaviors
- If we use https://github.com/Opentrons/opentrons/blob/5a224b0e4c4c0614ac1d5af2091b32219ec7579a/api/src/opentrons/hardware_control/types.py#L266 as an intermediate type, we can push the functionality of `do_firmware_updates` up to `ot3api`
- Also, I think this would help us avoid having `OT3Controller._attached_tools`, which is a duplicate cache of pretty complex functionality in `ot3api` - we should feed that information from the cache at the higher level

I do agree with pretty much everything you've done, just not necessarily _where_ you've done it - and that's just context with the codebase.

vegano1 committed:
Merge branch 'edge' into RCORE-477_drive_update-required

vegano1 committed:
fixing shit

vegano1 committed:
- added do_firmware_updates funtion to ot3Controller which is responsible for starting any firmware updates on the system
- start firmware updates on startup by calling do_firmware_updates from opentrons.__init__ initialize() function before homing the gantry
- move all of the checking for update logic to the opentrons_hardware module and just pass in the data we need from the ot3Controller
- remove get_update_type since we no longer need that and instead consolidate it into cehck_firmware_updates
- add update_required setter and message so its clear when we enter/exit firmware update mode
- change the way we determine the firmwareUpdateType for pipettes by using the number of channels mapped to the firmwareUpdateType
- changed the number of of bytes for FirmwareShortSHADataField from 7 to 8 to match the output of the generated shortsha as well as match what the devices keep in non-volatile.
- load_firmware_updates now gets called directly from check_firmware_updates so it just serializes the file on the fly instead of keeping it in memory
- handle BootloaderNotReady exception in firmware_update.run so we dont kill our asyncio update tasks
- changed the firmware_update.run.run_update function so it takes in a filepath and opens the file in the call instead of requring a TextIO hex file to be passed in.
- am now using run_updates function to do all the firmware updates concurrently as asyncio tasks
- run_updates update_details argument is now a Dict[NodeId, str] instead of Dict[NodeId, TextIO]
- convert the type argument for the update_fw.py script ffrom FileType to str since run_update now takes a filepath and opens the file itself.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +195,42 @@ def fw_version(self) -> Optional[str]:
+            # remove gripper from list of attached tools
+            node_id = sensor_node_for_mount(mount)
+            if node_id != NodeId.gripper:
+                pipettes[node_id] = pipette.get("channels", 0)
```
I think that rather than checking the channel count, which is an artifact of a higher level - it's data established and controlled by `ot3api` and meant for consumption largely by the protocol engine - we should massage the data in `ot3api` where this is called _from_ - so we just give it `pipette_kind: Dict[OT3Mount, PipetteType]` where `PipetteType` is from somewhere or other, maybe write it new

sfoster1 replied:
I don't think this was resolved - we're still checking the channel count directly here.

sfoster1 commented on hardware/opentrons_hardware/firmware_bindings/messages/fields.py:
```diff
@@ -31,7 +31,7 @@ class FirmwareShortSHADataField(utils.BinaryFieldBase[bytes]):
     """
 
-    NUM_BYTES = 7
+    NUM_BYTES = 8
```
Is this a change to the number of bytes on the wire, or were we incorrectly not parsing the 8th byte before?

vegano1 replied:
we were not parsing the 8th byte

sfoster1 reviewed (commented):
Looking really solid. I think `OT3Controller` is still doing a little too much, but overall good!

vegano1 committed:
- renamed do_update_firmware to firmware_update and added optional nodes arg so we can update specific nodes
- probe the network after updates are done so we can refresh our device info cache
- added files_by_revision to UpdateInfo to accomodate opentrons-firmware.json manifest change

vegano1 committed:
mypy

vegano1 committed:
Merge branch 'edge' into RCORE-477_drive_update-required

vegano1 committed:
gate firmware updates behind enableOT3FirmwareUpdates feature flag.

sfoster1 commented on api/src/opentrons/__init__.py:
```diff
@@ -159,13 +159,21 @@ async def _blink() -> None:
+            await hardware.do_firmware_updates()
+
     try:
+        if ff.enable_ot3_firmware_updates():
```
[nit]: let's push the feature flag check inside `_do_updates` in parallel with `should_use_ot3` so the high-level behavior of startup is more similar when the flag is or isn't enabled - create the task either way. It creates one less thing that can fail in one configuration but not the other.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +194,60 @@ def fw_version(self) -> Optional[str]:
+    async def update_firmware(
+        self,
+        attached_pipettes: Dict[OT3Mount, PipetteDict],
+        nodes: Optional[Set[NodeId]] = set(),
```
I think this should default `None` if it's going to be optional, unless empty-set and `None` have different behavior. If they do, we should still default this to none and change it if we ever modify this argument, because of the unsafety of binding objects as default arguments in python - `set()` will be evaluated at parse time, and the resulting instance stored on the class, so if this function is called with the default argument for `nodes` and it's modified, that modification will be present on subsequent calls too.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +194,60 @@ def fw_version(self) -> Optional[str]:
+        pipettes: Dict[NodeId, int] = dict()
+        for mount, pipette in attached_pipettes.items():
+            # remove gripper from list of attached tools
+            node_id = sensor_node_for_mount(mount)
```
let's not reuse `sensor_node_for_mount` directly - what if we end up wanting to change it in the future and use a different node id for sensors? let's try going from mount to subsystem and then subsystem to node id, should be more flexible.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +194,60 @@ def fw_version(self) -> Optional[str]:
+        for mount, pipette in attached_pipettes.items():
+            # remove gripper from list of attached tools
+            node_id = sensor_node_for_mount(mount)
+            if node_id != NodeId.gripper:
```
at that point we can make decisions like this based on subsystem rather than node ID which I think is a little nicer

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +194,60 @@ def fw_version(self) -> Optional[str]:
+                pipettes[node_id] = pipette.get("channels", 0)
+
+        # Force Update specified nodes, otherwise update everything which requires an update.
+        force = False
```
i'm always a little suspicious of function-scope variables like this that act as flags, I think it's a little easy to mess them up. What if instead we do something like
```
if nodes:
    force_device_info = {node: cache_info ...}
    firmware_updates = firmware_update.check_firmware_updates(force_device_info, pipettes, force=True)
else:
    firmware_updates = firmware_update.check_firmware_updates(self._network_info.device_info, pipettes, force=False)
```

we could even factor the above into its own little helper function, maybe even defined inside the body of `update_firmware`, so that we could eventually do `firmware_updates = get_firmware_updates(pipettes, nodes)`

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +194,60 @@ def fw_version(self) -> Optional[str]:
+        )
+
+        # Start firmware updates if we have any
+        if firmware_updates:
```
sort of a nit, but if we invert this:
```
if not firmware_updates:
    log.info("No firmware updates required")
    self.update_required = False
    return

log.info("Firmware updates are available.")
self.update_required = True
...
```
then it's a bit easier to read the rest of the function since you no longer have to keep in mind that you're in the scope of the `if`.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +194,60 @@ def fw_version(self) -> Optional[str]:
+
+    async def update_firmware(
+        self,
+        attached_pipettes: Dict[OT3Mount, PipetteDict],
```
I'm still a little uncomfortable about this being passed in. I think this data just doesn't belong in `ot3controller`. What if we have this function take a list of "peripheral subsystems to check"?

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3simulator.py:
```diff
@@ -461,9 +462,22 @@ def fw_version(self) -> Optional[str]:
+        nodes: Optional[Set[NodeId]] = set(),
+    ) -> None:
+        """Updates the firmware on the OT3."""
         pass
```
this probably wants a yield

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -400,13 +399,11 @@ async def create_simulating_module(
+    async def do_firmware_updates(self) -> None:
+        """Update all the firmware."""
+        # get the attached instruments so we can get the type of pipettes attached
+        attached_pipettes = self._pipette_handler.get_attached_instruments()
```
what if we transform this into a list of "attached tool subsystems" and pass _that_ into `update_firmware`?

sfoster1 commented on hardware/opentrons_hardware/firmware_update/utils.py:
```diff
@@ -0,0 +1,173 @@
+
+
+def load_firmware_manifest(
+    firmware_manifest: Union[str, Path] = str(),
```
we shouldn't use empty-string as a default here, both because it means the default variable could end up being mutated and because semantically it's unclear from the function signature. Instead, let's use `Union[str, Path, None]` or even better just pick `Optional[Path]` and `=None` to indicate that it's optional.

sfoster1 commented on hardware/opentrons_hardware/firmware_update/utils.py:
```diff
@@ -0,0 +1,173 @@
+    """Serializes the opentrons firmware update info."""
+    opentrons_manifest: Dict[FirmwareUpdateType, UpdateInfo] = dict()
+    opentrons_firmware = firmware_manifest or _FIRMWARE_MANIFEST_PATH
+    if not os.path.exists(opentrons_firmware):
```
[nit]: imo we should just immediately try and open it, and catch and reformat exceptions

sfoster1 commented on hardware/opentrons_hardware/firmware_update/utils.py:
```diff
@@ -0,0 +1,173 @@
+def check_firmware_updates(
+    device_info: Dict[NodeId, DeviceInfoCache],
+    attached_pipettes: Dict[NodeId, int],
+    force: Optional[bool] = False,
```
if we're going to have this be `Optional` it should default to `None`

sfoster1 commented on hardware/opentrons_hardware/firmware_update/utils.py:
```diff
@@ -0,0 +1,173 @@
+    for node, version_cache in device_info.items():
+        log.debug(f"Checking firmware update for {node.name}")
+        # Get the update type based on the channels if pipette
+        if attached_pipettes and node in [NodeId.pipette_left, NodeId.pipette_right]:
```
This is definitely too far down to push channel-count data and math. It's a concern for the hardware controller, and the hardware controller should be informing this function just what kind of pipette we're dealing with by ideally passing in a `FirmwareUpdateType` or something.

sfoster1 commented on hardware/opentrons_hardware/firmware_update/utils.py:
```diff
@@ -0,0 +1,173 @@
+            )
+            # TODO (BA, 02/03/2022): Get the filepath based on the revision once we add it to GetDeviceInfoResponse.
+            # For now just select the first value.
+            update_info.filepath = list(update_info.files_by_revision.values())[0]
```
I think this will cause problems with https://github.com/Opentrons/oe-core/pull/53 , let's maybe always pull `update_info.files_by_revision["rev1"]` instead

sfoster1 commented on hardware/opentrons_hardware/hardware_control/network.py:
```diff
@@ -54,7 +54,7 @@ def nodes(self) -> Set[NodeId]:
 
     async def probe(
-        self, expected: Optional[Set[NodeId]], timeout: float = 1.0
+        self, expected: Optional[Set[NodeId]] = set(), timeout: float = 1.0
```
if we want a default value here it should be `None` since this is `Optional`

sfoster1 reviewed (changes requested):
There's a couple smaller things I have in there about what the right way to present an optional is and what default arguments should be, but I think there's one other problem which is layer violations and intermingling of types.

In my opinion, `channel_count` shouldn't ever be exposed to `opentrons_hardware` and it shouldn't make decisions on it. Instead, those decisions should be made in the hardware controller with some mapping from pipette types or channels to `opentrons_hardware` firmware update details and sent down that way. Using channels implicitly binds in data about pipette functionality that `opentrons_hardware` otherwise doesn't need to know, and it's worth adding a couple more layers of indirection to keep `opentrons_hardware` as a leaf module that consumes no information from `opentrons`.

vegano1 committed:
- added PipetteType which gets passed in from ot3api to map channels to firmwareUpdateType
- moved pipette type generation logic from ot3Controller to ot3api
- removed force flag from update_firmware and instead using whether nodes are passed in to force update
= added firmware_updates/utils.py unit tests
- removed casting from _generate_firmware_info so we can capture any issues with exception
- only update specified node if nodes are given in check_for_updates
- added OT3Controller.update_firmware unit tests

sfoster1 commented on api/src/opentrons/hardware_control/ot3api.py:
```diff
@@ -400,13 +405,19 @@ async def create_simulating_module(
+                node_id = sensor_node_for_mount(mount)
+                channels = pipette.get("channels")
+                if channels:
+                    pipettes[node_id] = channel_to_pipette_type(channels)
```
Love this refactor, but one little thing is that we try and keep imports from `opentrons_hardware` specifically to `ot3controller`. This does make things a little painful, but the benefit is that it makes it a little easier to massage optional python dependencies. I think the appropriate interface here might be
- Make a new enum in `opentrons_hardware.types` called like `OT3UpdatableSubsystem` that has entries of like `gantry_x`, `gantry_y`, `head`, `pipette_single`, `pipette_multi`, `pipette_96`, `gripper`
- This function builds a `Dict[OT3Mount, OT3UpdatableSubsystem]` and passes it to `OT3Controller.update_firmware`
- `OT3Controller.update_firmware` transforms it into a `Dict[NodeId, PipetteType]`
This is obviously more code, more work, more layers of indirection, but what it _does_ do is cleanly separate the two. If we push hardware control stuff down at some point, or get `opentrons_hardware` onto the OT2, we can remove the redundancy since we won't need the split.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +195,39 @@ def fw_version(self) -> Optional[str]:
+
+    async def update_firmware(
+        self,
+        attached_pipettes: Dict[NodeId, PipetteType],
```
as mentioned in the comment on `ot3api`, I think the ideal case here is that this argument is specifically a `Dict[OT3Mount, OT3UpdatableSubsystem]` where `OT3UpdatableSubsystem` is pretty much the same thing as the device type enum in `opentrons_hardware`. That keeps the translation between "hardware controller concept" and "`opentrons_hardware` concept" directly in this file, and therefore they remain nice and separated.

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -194,19 +195,39 @@ def fw_version(self) -> Optional[str]:
+    ) -> None:
+        """Updates the firmware on the OT3."""
+        # Check if devices need an update, force update if nodes are specified
+        firmware_updates = firmware_update.check_firmware_updates(
```
we'd insert a transformation before this line to turn the `Dict[OT3Mount, OT3UpdatableSubsystem]` into `Dict[NodeId, PipetteType]` here

sfoster1 reviewed (changes requested):
Love this refactor but I think we can tease the hardware control and opentrons_hardware layers apart just slightly farther

vegano1 committed:
separate opentrons_hardware package a bit more by removing concept of channels when determining pipette type and instead use enum to trickle that data down from opentrons package.

vegano1 committed:
Merge branch 'edge' into RCORE-477_drive_update-required

vegano1 committed:
integrate firmware update status

sfoster1 commented on api/src/opentrons/hardware_control/backends/ot3controller.py:
```diff
@@ -619,9 +658,9 @@ async def get_attached_instruments(
+        attached_tools = await self._tool_detector.detect()
 
-        current_tools = dict(OT3Controller._generate_attached_instrs(attached))
+        current_tools = dict(OT3Controller._generate_attached_instrs(attached_tools))
```
nit: returns a dict already right

vegano1 replied:
weird looks like I just renamed this to attached_tools while testing, reverting this back to edge.
this is converting a tuple of [OT3Mount, OT3AttachedInstruments] to Dict.

sfoster1 commented on hardware/manifest.json:
```diff
@@ -0,0 +1 @@
+{"manifest_version": 1, "subsystems": {"head": {"version": 2, "shortsha": "25755efd", "files_by_revision": {"rev1": "head-rev1.hex"}}}}
```
is this test data? if so it should go in the test area.

vegano1 replied:
local testing, removed

vegano1 committed:
remove test files

sfoster1 commented on hardware/opentrons_hardware/firmware_bindings/constants.py:
```diff
@@ -228,6 +228,16 @@ class PipetteName(int, Enum):
+    pipette_single = 1
+    pipette_multi = 2
+    pipette_96 = 3
+    pipette_384 = 4
```
we should pull this one for now - it's too much future proofing

sfoster1 commented on hardware/opentrons_hardware/firmware_update/run.py:
```diff
@@ -62,69 +62,88 @@ async def _run_update(
         """Perform a firmware update on a node target."""
-        hex_processor = HexRecordProcessor.from_file(hex_file)
+        if not os.path.exists(filepath):
+            print("Subsystem update file not found {filepath}")
```
should be a `log.error`

sfoster1 commented on hardware/opentrons_hardware/firmware_update/run.py:
```diff
@@ -62,69 +62,88 @@ async def _run_update(
 
-        initiator = FirmwareUpdateInitiator(messenger)
-        downloader = FirmwareUpdateDownloader(messenger)
+        with open(filepath) as f:
```
similar to the thing with doing `if not <x>: return` to make a function easier to read, can we either factor the content that's inside the `with` into a separate function that takes a `fileio`, or change this back to taking an opened file?

sfoster1 commented on hardware/opentrons_hardware/firmware_update/utils.py:
```diff
@@ -0,0 +1,187 @@
+    pipettes_single = 4
+    pipettes_multi = 5
+    pipettes_96 = 6
+    pipettes_384 = 7
```
let's pull this out for now

sfoster1 reviewed (changes requested):
This is looking awesome! Couple small changes but definitely almost there

vegano1 committed:
remove pipette-384
smal changes

sfoster1 commented on api/tests/opentrons/hardware_control/backends/test_ot3_controller.py:
```diff
@@ -805,3 +810,124 @@ async def fake_umpe(
             assert False, "update_motor_estimation raised an exception."
+
+
+async def test_update_firmware(controller: OT3Controller) -> None:
```
I think really big tests like this that cover multiple cases are better implemented as multiple smaller tests that each cover one case. That means that something that causes all those cases to fail can be detected as such, rather than fixing one case -> on to the next one and now _that_ one's failing, so on and so forth

sfoster1 commented on api/tests/opentrons/hardware_control/backends/test_ot3_controller.py:
```diff
@@ -790,10 +792,13 @@ async def fake_umpe(
+            try:
+                await controller.update_firmware({})
+            except FirmwareUpdateRequired:
+                assert False, "update_firmware raised an exception."
```
won't this already be raising an exception? why the catch and assert?

vegano1 replied:
This specific method wont raise an exception, this is just testing that.

sfoster1 commented on hardware/tests/opentrons_hardware/firmware_update/test_utils.py:
```diff
@@ -0,0 +1,226 @@
+    with open(manifest_filename, "w") as fp:
+        json.dump(mock_manifest, fp)
+
+    # test that the file written to disk can be serialized
```
```suggestion
    # test that the file written to disk can be deserialized
```
(nit)

sfoster1 commented on hardware/tests/opentrons_hardware/firmware_update/test_utils.py:
```diff
@@ -0,0 +1,226 @@
+        os.unlink(manifest_filename)
+
+
+def test_load_firmware_manifest_failure(mock_manifest: Dict[str, Any]) -> None:
```
this is another of those really big tests that test multiple things. we should refactor that into multiple smaller tests so one test failing doesn't block others. in general, these tests where you have multiple blocks of code each with a comment that says what it tests can and should be split at the comments.

sfoster1 commented on hardware/tests/opentrons_hardware/firmware_update/test_utils.py:
```diff
@@ -0,0 +1,226 @@
+    os.unlink(manifest_filename)
+
+
+def test_check_firmware_updates_available(mock_manifest: Dict[str, Any]) -> None:
```
split this

sfoster1 reviewed (changes requested):
Couple testing things

vegano1 committed:
split up tests

sfoster1 reviewed (approved):
Looks excellent to me once lint passes!

vegano1 committed:
mypy
